src/modules/dataFetching/components/GenericTableWithData.tsx

```diff
@@ -149,15 +142,17 @@ const GenericTableWithData = <
   const previousQueryVariables = usePrevious(queryVariables);
   const [filterValues, setFilterValues] = useState(defaultFilterValues);
   const [sortOrder, setSortOrder] = useState<typeof defaultSortOptionProp>();
-  const [includedOptionalColumns, setIncludedOptionalColumns] = useState<
-    string[]
-  >(
-    compact(
-      columnsProp
-        ?.filter((col) => col.optional && !col.defaultHidden)
-        .map((col) => col.key) || []
-    )
-  );
+
+  // Careful: Optional column behavior is undefined/unsupported when columns are provided by
+  // getColumnDefs instead of the columns prop. This is because `getColumnDefs` receives `rows`,
+  // but `rows` comes from the graphql query that receives `optionalQueryVariables`,
+  // leading to a circular dependency.
+  const {
+    optionalColumns,
+    includedOptionalColumns,
+    setIncludedOptionalColumns,
+    optionalQueryVariables,
+  } = useOptionalColumns<RowDataType, QueryVariables>({ columns: columnsProp });
```

Comment:
I had one more follow up on this while reviewing  https://github.com/greenriver/hmis-frontend/pull/1076:

I'm not sure I would understand what (not) to do as a developer when reading this comment. It seems like this hook should just do nothing if `columnsProp` is missing, right?

What happens if you pass a `getColumnDefs` function that returns a column that is specified as optional? Looking at the code down on line 271 I think optional columns would just always be hidden, is that right?

Reply:
It looks like `getColumnDefs` is only used in a handful of cases. Maybe we should look into replacing it, or at least capture future work of how to make it function with optional cols. Looking at the usages:
- BulkServicesTable - used in order to get loading prop. maybe that can be passed to the row render function instead
- EnrollmentCaseNotes - used in order to conditionally include CDE columns based on whether the page has any records with CDEs
- EnrollmentCLS - same as case notes usage
- ProjectExternalSubmissionsTable - similar to CDE use-case

Reply:
Thanks for looking into `getColumnDefs` usage, and sorry my comment was confusing! I can update it to be a little more clear, but yeah basically the "what to do as a developer" I had in mind when writing this comment was, "if you find yourself needing to use `getColumnDefs` with optional columns, it probably points to the necessity for refactoring `getColumnDefs`." I can open a tech debt ticket so we don't lose track of this.

Reply:
Created https://github.com/open-path/Green-River/issues/7387